diff --git a/serialization/_torch.py b/serialization/_torch.py
index 2bad2fc..c5c70fc 100644
--- a/serialization/_torch.py
+++ b/serialization/_torch.py
@@ -783,7 +783,7 @@ def get_torch_storage_size(tensor: "torch.Tensor") -> int:
         pass

     try:
-        return tensor.untyped_storage().nbytes()
+        return tensor.nbytes
     except AttributeError:
         # Fallback for torch==1.10
         try:
@@ -830,7 +830,7 @@ def storage_ptr(tensor: "torch.Tensor") -> Union[int, Tuple[Any, ...]]:
         pass

     try:
-        return tensor.untyped_storage().data_ptr()
+        return tensor
     except Exception:
         # Fallback for torch==1.10
         try:
