diff --git a/pretrain_gpt.py b/pretrain_gpt.py
index 8c234aaf..e2e87e68 100644
--- a/pretrain_gpt.py
+++ b/pretrain_gpt.py
@@ -3,6 +3,7 @@
 
 import os
 import torch
+import ascendspeed.megatron_adaptor
 from functools import partial
 from typing import Union
 from megatron.training import get_args
@@ -77,8 +78,6 @@ def model_provider(pre_process=True, post_process=True) -> Union[GPTModel, megat
             rotary_percent=args.rotary_percent,
         )
     else:
-        assert(args.context_parallel_size == 1), "Context parallelism is only supported with Megatron Core!"
-
         model = megatron.legacy.model.GPTModel(
             config,
             num_tokentypes=0,
